File: server/libs/physics/rigidbody.rs
use std::default;

use crate::vec::Vec3;

use super::aabb::AABB;

pub struct RigidBody {
    air_drag: f32,
    fluid_drag: f32,

    pub resting: Vec3<i32>,
    pub velocity: Vec3<f32>,
    pub in_fluid: bool,
    pub ratio_in_fluid: f32,
    pub forces: Vec3<f32>,
    pub impulses: Vec3<f32>,
    pub sleep_frame_count: u32,

    pub aabb: AABB,
    pub mass: f32,
    pub friction: f32,
    pub restitution: f32,
    pub gravity_multiplier: f32,
    pub auto_step: bool,
}

impl RigidBody {}

#[derive(Default)]
pub struct RigidBodyBuilder {
    aabb: AABB,
    mass: f32,
    friction: f32,
    restitution: f32,
    gravity_multiplier: f32,
    auto_step: bool,
}

impl RigidBodyBuilder {
    pub fn new(aabb: &AABB) -> Self {
        Self {
            aabb: aabb.to_owned(),
            ..Default::default()
        }
    }

    pub fn mass(mut self, mass: f32) -> Self {
        self.mass = mass;
        self
    }

    pub fn friction(mut self, friction: f32) -> Self {
        self.friction = friction;
        self
    }

    pub fn restitution(mut self, restitution: f32) -> Self {
        self.restitution = restitution;
        self
    }

    pub fn gravity_multiplier(mut self, gravity_multiplier: f32) -> Self {
        self.gravity_multiplier = gravity_multiplier;
        self
    }

    pub fn auto_step(mut self, auto_step: bool) -> Self {
        self.auto_step = auto_step;
        self
    }

    pub fn build(self) -> RigidBody {
        RigidBody {
            air_drag: -1.0,
            fluid_drag: -1.0,

            resting: Vec3::default(),
            velocity: Vec3::default(),
            in_fluid: false,
            ratio_in_fluid: 0.0,
            forces: Vec3::default(),
            impulses: Vec3::default(),
            sleep_frame_count: 10,

            aabb: self.aabb,
            mass: self.mass,
            friction: self.friction,
            restitution: self.restitution,
            gravity_multiplier: self.gravity_multiplier,
            auto_step: self.auto_step,
        }
    }
}

File: packages/physics-engine/src/rigid-body.ts
import { AABB } from '@voxelize/voxel-aabb';

// massive thanks to https://github.com/andyhall/voxel-physics-engine/blob/master/src/rigidBody.js

class RigidBody {
  public airDrag: number;
  public fluidDrag: number;

  public resting = [0, 0, 0];
  public velocity = [0, 0, 0];
  public inFluid = false;
  public ratioInFluid = 0;
  public forces = [0, 0, 0];
  public impulses = [0, 0, 0];
  public sleepFrameCount = 10 | 0;

  constructor(
    public aabb: AABB,
    public mass: number,
    public friction: number,
    public restitution: number,
    public gravityMultiplier: number,
    public stepHeight: number,
    public onStep?: (newAABB: AABB, resting: number[]) => void,
    public onCollide?: (impacts?: number[]) => void,
  ) {
    this.airDrag = -1;
    this.fluidDrag = -1;
  }

  setPosition = (p: number[]) => {
    this.aabb.setPosition([
      p[0] - this.aabb.width / 2,
      p[1] - this.aabb.height / 2,
      p[2] - this.aabb.depth / 2,
    ]);

    this.markActive();
  };

  getPosition = () => {
    return [
      this.aabb.minX + this.aabb.width / 2,
      this.aabb.minY + this.aabb.height / 2,
      this.aabb.minZ + this.aabb.depth / 2,
    ];
  };

  applyForce = (f: number[]) => {
    this.forces[0] += f[0];
    this.forces[1] += f[1];
    this.forces[2] += f[2];
    this.markActive();
  };

  applyImpulse = (i: number[]) => {
    this.impulses[0] += i[0];
    this.impulses[1] += i[1];
    this.impulses[2] += i[2];
    this.markActive();
  };

  markActive = () => {
    this.sleepFrameCount = 10 | 0;
  };

  get atRestX() {
    return this.resting[0];
  }

  get atRestY() {
    return this.resting[1];
  }

  get atRestZ() {
    return this.resting[2];
  }
}

export { RigidBody };
